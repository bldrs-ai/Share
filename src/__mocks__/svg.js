/* eslint-disable max-len */
/* eslint-disable no-magic-numbers */
export const mockLoadedSvgObj = {
  paths: [
    {
      type: 'ShapePath',
      color: 0,
      subPaths: [
        {
          metadata: {
            version: 4.5,
            type: 'Curve',
            generator: 'Curve.toJSON',
          },
          arcLengthDivisions: 200,
          type: 'Path',
          autoClose: true,
          curves: [
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                349.82,
                44.801,
              ],
              v1: [
                272.496,
                44.801,
              ],
              v2: [
                210,
                107.375,
              ],
              v3: [
                210,
                184.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                210,
                184.801,
              ],
              v1: [
                210,
                202.582,
              ],
              v2: [
                215.184,
                225.164,
              ],
              v3: [
                221.199,
                240.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'LineCurve',
              v1: [
                221.199,
                240.801,
              ],
              v2: [
                349.999,
                515.201,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'LineCurve',
              v1: [
                349.999,
                515.201,
              ],
              v2: [
                478.799,
                240.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                478.799,
                240.801,
              ],
              v1: [
                484.846,
                225.156,
              ],
              v2: [
                489.998,
                202.59,
              ],
              v3: [
                489.998,
                184.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                489.998,
                184.801,
              ],
              v1: [
                489.998,
                107.375,
              ],
              v2: [
                427.15,
                44.801,
              ],
              v3: [
                349.818,
                44.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'LineCurve',
              v1: [
                349.818,
                44.801,
              ],
              v2: [
                349.82,
                44.801,
              ],
            },
          ],
          currentPoint: [
            349.82,
            44.801,
          ],
        },
        {
          metadata: {
            version: 4.5,
            type: 'Curve',
            generator: 'Curve.toJSON',
          },
          arcLengthDivisions: 200,
          type: 'Path',
          autoClose: true,
          curves: [
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                349.996,
                117.602,
              ],
              v1: [
                387.113,
                117.602,
              ],
              v2: [
                417.195,
                147.637,
              ],
              v3: [
                417.195,
                184.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                417.195,
                184.801,
              ],
              v1: [
                417.195,
                221.965,
              ],
              v2: [
                387.113,
                252,
              ],
              v3: [
                349.996,
                252,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                349.996,
                252,
              ],
              v1: [
                312.879,
                252,
              ],
              v2: [
                282.797,
                221.965,
              ],
              v3: [
                282.797,
                184.801,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'CubicBezierCurve',
              v0: [
                282.797,
                184.801,
              ],
              v1: [
                282.797,
                147.637,
              ],
              v2: [
                312.879,
                117.602,
              ],
              v3: [
                349.996,
                117.602,
              ],
            },
            {
              metadata: {
                version: 4.5,
                type: 'Curve',
                generator: 'Curve.toJSON',
              },
              arcLengthDivisions: 200,
              type: 'LineCurve',
              v1: [
                349.996,
                117.602,
              ],
              v2: [
                349.996,
                117.602,
              ],
            },
          ],
          currentPoint: [
            349.996,
            117.602,
          ],
        },
      ],
      currentPath: {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'Path',
        autoClose: true,
        curves: [
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              349.996,
              117.602,
            ],
            v1: [
              387.113,
              117.602,
            ],
            v2: [
              417.195,
              147.637,
            ],
            v3: [
              417.195,
              184.801,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              417.195,
              184.801,
            ],
            v1: [
              417.195,
              221.965,
            ],
            v2: [
              387.113,
              252,
            ],
            v3: [
              349.996,
              252,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              349.996,
              252,
            ],
            v1: [
              312.879,
              252,
            ],
            v2: [
              282.797,
              221.965,
            ],
            v3: [
              282.797,
              184.801,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              282.797,
              184.801,
            ],
            v1: [
              282.797,
              147.637,
            ],
            v2: [
              312.879,
              117.602,
            ],
            v3: [
              349.996,
              117.602,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'LineCurve',
            v1: [
              349.996,
              117.602,
            ],
            v2: [
              349.996,
              117.602,
            ],
          },
        ],
        currentPoint: [
          349.996,
          117.602,
        ],
      },
      userData: {
        node: {},
        style: {
          fill: '#000',
          fillOpacity: 1,
          strokeOpacity: 1,
          strokeWidth: 1,
          strokeLineJoin: 'miter',
          strokeLineCap: 'butt',
          strokeMiterLimit: 4,
        },
      },
    },
  ],
}

export const mockShapes = [
  {
    metadata: {
      version: 4.5,
      type: 'Curve',
      generator: 'Curve.toJSON',
    },
    arcLengthDivisions: 200,
    type: 'Shape',
    autoClose: false,
    curves: [
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'CubicBezierCurve',
        v0: [
          349.82,
          44.801,
        ],
        v1: [
          272.496,
          44.801,
        ],
        v2: [
          210,
          107.375,
        ],
        v3: [
          210,
          184.801,
        ],
      },
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'CubicBezierCurve',
        v0: [
          210,
          184.801,
        ],
        v1: [
          210,
          202.582,
        ],
        v2: [
          215.184,
          225.164,
        ],
        v3: [
          221.199,
          240.801,
        ],
      },
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'LineCurve',
        v1: [
          221.199,
          240.801,
        ],
        v2: [
          349.999,
          515.201,
        ],
      },
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'LineCurve',
        v1: [
          349.999,
          515.201,
        ],
        v2: [
          478.799,
          240.801,
        ],
      },
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'CubicBezierCurve',
        v0: [
          478.799,
          240.801,
        ],
        v1: [
          484.846,
          225.156,
        ],
        v2: [
          489.998,
          202.59,
        ],
        v3: [
          489.998,
          184.801,
        ],
      },
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'CubicBezierCurve',
        v0: [
          489.998,
          184.801,
        ],
        v1: [
          489.998,
          107.375,
        ],
        v2: [
          427.15,
          44.801,
        ],
        v3: [
          349.818,
          44.801,
        ],
      },
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'LineCurve',
        v1: [
          349.818,
          44.801,
        ],
        v2: [
          349.82,
          44.801,
        ],
      },
    ],
    currentPoint: [
      0,
      0,
    ],
    uuid: '07E2D9F4-36B9-4A1E-8890-FACE42A7F871',
    holes: [
      {
        metadata: {
          version: 4.5,
          type: 'Curve',
          generator: 'Curve.toJSON',
        },
        arcLengthDivisions: 200,
        type: 'Path',
        autoClose: false,
        curves: [
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              349.996,
              117.602,
            ],
            v1: [
              387.113,
              117.602,
            ],
            v2: [
              417.195,
              147.637,
            ],
            v3: [
              417.195,
              184.801,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              417.195,
              184.801,
            ],
            v1: [
              417.195,
              221.965,
            ],
            v2: [
              387.113,
              252,
            ],
            v3: [
              349.996,
              252,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              349.996,
              252,
            ],
            v1: [
              312.879,
              252,
            ],
            v2: [
              282.797,
              221.965,
            ],
            v3: [
              282.797,
              184.801,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'CubicBezierCurve',
            v0: [
              282.797,
              184.801,
            ],
            v1: [
              282.797,
              147.637,
            ],
            v2: [
              312.879,
              117.602,
            ],
            v3: [
              349.996,
              117.602,
            ],
          },
          {
            metadata: {
              version: 4.5,
              type: 'Curve',
              generator: 'Curve.toJSON',
            },
            arcLengthDivisions: 200,
            type: 'LineCurve',
            v1: [
              349.996,
              117.602,
            ],
            v2: [
              349.996,
              117.602,
            ],
          },
        ],
        currentPoint: [
          0,
          0,
        ],
      },
    ],
  },
]

export const mockLoadedSvgStr = `<?xml version="1.0" encoding="utf-8"?>
<svg viewBox="210 44.801 279.998 470.4" width="279.998pt" height="470.4pt" xmlns="http://www.w3.org/2000/svg">
  <path d="M 349.82 44.801 C 272.496 44.801 210 107.375 210 184.801 C 210 202.582 215.184 225.164 221.199 240.801 L 349.999 515.201 L 478.799 240.801 C 484.846 225.156 489.998 202.59 489.998 184.801 C 489.998 107.375 427.15 44.801 349.818 44.801 L 349.82 44.801 Z M 349.996 117.602 C 387.113 117.602 417.195 147.637 417.195 184.801 C 417.195 221.965 387.113 252 349.996 252 C 312.879 252 282.797 221.965 282.797 184.801 C 282.797 147.637 312.879 117.602 349.996 117.602 L 349.996 117.602 Z"/>
</svg>`
