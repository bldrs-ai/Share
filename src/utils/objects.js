/**
 * Delete all properties defined in the given object.
 * @param {Object} obj The object whose properties to delete.
 */
export function deleteProperties(obj) {
  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      delete obj[key]
    }
  }
}


/**
 * @param {Object} obj
 * @return {boolean} True iff val is an object
 */
export function isObject(obj) {
  // Search for Daan in:
  // https://stackoverflow.com/questions/8511281/check-if-a-value-is-an-object-in-javascript
  return obj === Object(obj) && Object.prototype.toString.call(obj) !== '[object Array]'
}
