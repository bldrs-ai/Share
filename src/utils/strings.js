/**
 * Convert string to integer.
 * @param {string} s Input string.
 * @return {Number} integer.
 */
export function stoi(s) {
  const i = parseInt(s)
  if (!isFinite(i)) {
    throw new Error('Expected integer, got: ' + s)
  }
  return i
}


/**
 * Create a simple key by removing any non alpha-numeric character
 * @param {string} str To convert
 * @return {string} The converted result
 */
export function toKey(str) {
  return str.replace(/[^a-zA-Z0-9]+/, '')
}
