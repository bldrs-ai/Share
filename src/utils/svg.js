import {
  FileLoader,
  Group,
  MeshBasicMaterial,
  DoubleSide,
  ShapeGeometry,
  Mesh,
  Box3,
  Texture,
  LinearFilter,
  SpriteMaterial,
  Sprite,
  ImageLoader,
  Color,
} from 'three'
import {SVGLoader} from 'three/examples/jsm/loaders/SVGLoader'
import {assertDefined} from './assert'


const svgLoader = new SVGLoader()
const fileLoader = new FileLoader()
const imageLoader = new ImageLoader()


/**
 * Get svg object from url
 *
 * @param {string} svgUrl
 * @return {object}
 */
export async function getSvgObjFromUrl(svgUrl) {
  const svgObj = await svgLoader.loadAsync(svgUrl)
  return svgObj
}


/**
 * Generate group using svg file
 *
 * @param {object} svgObj
 * @param {string} fillColor color to fill group
 * @param {string} strokeColor color to draw strokes
 * @param {number} width
 * @param {number} height
 * @param {boolean} drawStrokes
 * @param {boolean} drawFillShapes
 * @param {boolean} strokesWireframe
 * @param {boolean} fillShapesWireframe
 * @return {number} svg based group
 */
export function getSvgGroupFromObj({
  svgObj,
  fillColor,
  strokeColor,
  width = 0,
  height = 2,
  drawStrokes = true,
  drawFillShapes = true,
  strokesWireframe = false,
  fillShapesWireframe = false,
}) {
  const paths = svgObj.paths
  const group = new Group()
  const svgGroup = new Group()
  const tempColor = new Color()

  for (let i = 0; i < paths.length; i++) {
    const path = paths[i]
    if (!fillColor) {
      fillColor = path.userData.style.fill
    }
    if (!strokeColor) {
      strokeColor = path.userData.style.stroke
    }

    if (drawFillShapes && fillColor !== undefined && fillColor !== 'none') {
      const color = tempColor.setStyle(fillColor).convertSRGBToLinear()
      const material = new MeshBasicMaterial({
        color: color,
        opacity: path.userData.style.fillOpacity,
        transparent: true,
        side: DoubleSide,
        // depthWrite: false,
        wireframe: fillShapesWireframe,
      })
      const shapes = SVGLoader.createShapes(path)

      for (let j = 0; j < shapes.length; j++) {
        const shape = shapes[j]
        const geometry = new ShapeGeometry(shape)
        const mesh = new Mesh(geometry, material)
        group.add(mesh)
      }
    }

    if (drawStrokes && strokeColor !== undefined && strokeColor !== 'none') {
      const color = tempColor.setStyle(strokeColor).convertSRGBToLinear()
      const material = new MeshBasicMaterial({
        color: color,
        opacity: path.userData.style.strokeOpacity,
        transparent: true,
        side: DoubleSide,
        // depthWrite: false,
        wireframe: strokesWireframe,
      })

      for (let j = 0, jl = path.subPaths.length; j < jl; j++) {
        const subPath = path.subPaths[j]
        const geometry = SVGLoader.pointsToStroke(subPath.getPoints(), path.userData.style)

        if (geometry) {
          const mesh = new Mesh(geometry, material)
          group.add(mesh)
        }
      }
    }
  }

  const groupBox3 = new Box3()
  groupBox3.setFromObject(group)
  const groupSize = groupBox3.max.sub(groupBox3.min)
  let scaleX = 0
  let scaleY = 0
  if (width) {
    scaleX = width / groupSize.x
  }
  if (height) {
    scaleY = height / groupSize.y
  }
  if (!width) {
    scaleX = scaleY
  }
  if (!height) {
    scaleY = scaleX
  }
  if (!width && !height) {
    scaleX = scaleY = 1
  }
  if (!width) {
    width = scaleX * groupSize.x
  }
  if (!height) {
    height = scaleY * groupSize.y
  }
  group.scale.x = scaleX
  group.scale.y = scaleY
  group.position.x = -width
  group.position.y = height
  group.scale.y *= - 1
  svgGroup.add(group)
  return svgGroup
}


/**
 * Get svg canvas
 *
 * @param {string} svgUrl
 * @param {string} fillColor
 * @return {object}
 */
export async function getSvgCanvas({svgUrl, fillColor}) {
  const svgStr = await fileLoader.loadAsync(svgUrl)
  const parser = new DOMParser()
  const svgEl = parser.parseFromString(svgStr, 'image/svg+xml').documentElement
  if (fillColor) {
    svgEl.setAttribute('fill', fillColor)
  }
  const newSvgData = (new XMLSerializer()).serializeToString(svgEl)
  const dataUrl = `data:image/svg+xml;base64,${window.btoa(unescape(encodeURIComponent(newSvgData)))}`
  const imageTag = await imageLoader.loadAsync(dataUrl)
  const canvas = document.createElement('canvas')
  canvas.width = svgEl.width?.baseVal?.value
  canvas.height = svgEl.height?.baseVal?.value
  const ctx = canvas.getContext('2d')
  ctx.drawImage(imageTag, 0, 0)
  return canvas
}


/**
 * Generate sprite using svg file
 *
 * @param {string} svgEl
 * @param {string} fillColor color to fill sprite
 * @param {number} width
 * @param {number} height
 * @return {number} svg based sprite
 */
export function getSpriteFromSvgCanvas({
  svgCanvas,
  width = 0,
  height = 0,
}) {
  assertDefined(svgCanvas)
  if (width <= 0) {
    width = height
  }
  if (height <= 0) {
    height = width
  }
  if (width <= 0 && height <= 0) {
    width = height = 1
  }
  const texture = new Texture(svgCanvas)
  texture.needsUpdate = true
  const material = new SpriteMaterial({map: texture, side: DoubleSide})
  material.map.minFilter = LinearFilter
  const sprite = new Sprite(material)
  sprite.scale.set(width, height, 1.0)
  return sprite
}


/**
 * Generate sprite using svg file
 *
 * @param {Group} group
 * @param {object} userData color to fill sprite
 */
export function addUserDataInGroup(group, userData) {
  assertDefined(group, userData)
  group.traverse((child) => {
    if (child instanceof Mesh) {
      child.userData = {
        ...child.userData,
        ...userData,
      }
    }
  })
  group.userData = {
    ...group.userData,
    ...userData,
  }
}


/**
 * @param {Group} group
 * @param {string} color
 */
export function setGroupColor(group, color) {
  assertDefined(group, color)
  const tempColor = new Color()
  group.traverse((child) => {
    if (child instanceof Mesh) {
      child.material.color = tempColor.set(color)
    }
  })
}

/**
 * @param {Group} group
 */
export function disposeGroup(group) {
  assertDefined(group)

  group.traverse((child) => {
    if (child instanceof Mesh) {
      child.geometry.dispose()
      child.material.dispose()
    }
  })
}
