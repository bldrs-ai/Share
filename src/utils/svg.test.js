import {deletePropertyRecursive, deleteStringValueMatchRecursive} from './objects'
import {UUID_REGEX} from './strings'
import {getSvgGroupFromObj, getSpriteFromSvgCanvas} from './svg'


const {SVGLoader} = jest.requireActual('three/examples/jsm/loaders/SVGLoader')


describe('svg', () => {
  it('getSvgGroupFromObj snapshot of PlaceMarkSvg matches', () => {
    const fs = require('fs')
    const svgBuf = fs.readFileSync('src/assets/icons/PlaceMark.svg', 'utf8')
    const svgLoader = new SVGLoader()
    const svgObj = svgLoader.parse(svgBuf)
    const svgGroup = getSvgGroupFromObj({svgObj})
    deletePropertyRecursive(svgGroup, 'uuid')
    deleteStringValueMatchRecursive(svgGroup, UUID_REGEX)
    expect(svgGroup).toMatchSnapshot()
  })

  it('getSpriteFromSvgCanvas snapshot of PlaceMarkSvg matches', () => {
    const svgCanvas = document.createElement('canvas')
    const svgSprite = getSpriteFromSvgCanvas({svgCanvas})
    deletePropertyRecursive(svgSprite, 'uuid')
    deletePropertyRecursive(svgSprite, 'images')
    deletePropertyRecursive(svgSprite, 'image')
    deleteStringValueMatchRecursive(svgSprite, UUID_REGEX)
    expect(svgSprite).toMatchSnapshot()
  })
})
