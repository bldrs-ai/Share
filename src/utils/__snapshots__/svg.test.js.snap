// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svg getSpriteFromSvgCanvas snapshot of PlaceMarkSvg matches 1`] = `
{
  "materials": [
    {
      "color": 16777215,
      "colorWrite": true,
      "depthFunc": 3,
      "depthTest": true,
      "depthWrite": true,
      "map": undefined,
      "side": 2,
      "sizeAttenuation": true,
      "stencilFail": 7680,
      "stencilFunc": 519,
      "stencilFuncMask": 255,
      "stencilRef": 0,
      "stencilWrite": false,
      "stencilWriteMask": 255,
      "stencilZFail": 7680,
      "stencilZPass": 7680,
      "transparent": true,
      "type": "SpriteMaterial",
      "uuid": undefined,
      "wireframeLinecap": undefined,
      "wireframeLinejoin": undefined,
    },
  ],
  "metadata": {
    "generator": "Object3D.toJSON",
    "type": "Object",
    "version": 4.5,
  },
  "object": {
    "layers": 1,
    "material": undefined,
    "matrix": [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
    ],
    "type": "Sprite",
    "uuid": undefined,
  },
  "textures": [
    {
      "anisotropy": 1,
      "center": [
        0,
        0,
      ],
      "encoding": 3000,
      "flipY": true,
      "format": 1023,
      "magFilter": 1006,
      "mapping": 300,
      "minFilter": 1006,
      "name": "",
      "offset": [
        0,
        0,
      ],
      "premultiplyAlpha": false,
      "repeat": [
        1,
        1,
      ],
      "rotation": 0,
      "type": 1009,
      "unpackAlignment": 4,
      "uuid": undefined,
      "wrap": [
        1001,
        1001,
      ],
    },
  ],
}
`;

exports[`svg getSvgGroupFromObj snapshot of PlaceMarkSvg matches 1`] = `
{
  "geometries": [
    {
      "curveSegments": 12,
      "shapes": [
        undefined,
      ],
      "type": "ShapeGeometry",
      "uuid": undefined,
    },
  ],
  "materials": [
    {
      "color": 0,
      "colorWrite": true,
      "depthFunc": 3,
      "depthTest": true,
      "depthWrite": true,
      "reflectivity": 1,
      "refractionRatio": 0.98,
      "side": 2,
      "stencilFail": 7680,
      "stencilFunc": 519,
      "stencilFuncMask": 255,
      "stencilRef": 0,
      "stencilWrite": false,
      "stencilWriteMask": 255,
      "stencilZFail": 7680,
      "stencilZPass": 7680,
      "transparent": true,
      "type": "MeshBasicMaterial",
      "uuid": undefined,
    },
  ],
  "metadata": {
    "generator": "Object3D.toJSON",
    "type": "Object",
    "version": 4.5,
  },
  "object": {
    "children": [
      {
        "children": [
          {
            "geometry": undefined,
            "layers": 1,
            "material": undefined,
            "matrix": [
              1,
              0,
              0,
              0,
              0,
              1,
              0,
              0,
              0,
              0,
              1,
              0,
              0,
              0,
              0,
              1,
            ],
            "type": "Mesh",
            "uuid": undefined,
          },
        ],
        "layers": 1,
        "matrix": [
          1,
          0,
          0,
          0,
          0,
          1,
          0,
          0,
          0,
          0,
          1,
          0,
          0,
          0,
          0,
          1,
        ],
        "type": "Group",
        "uuid": undefined,
      },
    ],
    "layers": 1,
    "matrix": [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
    ],
    "type": "Group",
    "uuid": undefined,
  },
  "shapes": [
    {
      "arcLengthDivisions": 200,
      "autoClose": false,
      "currentPoint": [
        0,
        0,
      ],
      "curves": [
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "CubicBezierCurve",
          "v0": [
            349.82,
            44.801,
          ],
          "v1": [
            272.496,
            44.801,
          ],
          "v2": [
            210,
            107.375,
          ],
          "v3": [
            210,
            184.801,
          ],
        },
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "CubicBezierCurve",
          "v0": [
            210,
            184.801,
          ],
          "v1": [
            210,
            202.582,
          ],
          "v2": [
            215.184,
            225.164,
          ],
          "v3": [
            221.199,
            240.801,
          ],
        },
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "LineCurve",
          "v1": [
            221.199,
            240.801,
          ],
          "v2": [
            349.999,
            515.201,
          ],
        },
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "LineCurve",
          "v1": [
            349.999,
            515.201,
          ],
          "v2": [
            478.799,
            240.801,
          ],
        },
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "CubicBezierCurve",
          "v0": [
            478.799,
            240.801,
          ],
          "v1": [
            484.846,
            225.156,
          ],
          "v2": [
            489.998,
            202.59,
          ],
          "v3": [
            489.998,
            184.801,
          ],
        },
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "CubicBezierCurve",
          "v0": [
            489.998,
            184.801,
          ],
          "v1": [
            489.998,
            107.375,
          ],
          "v2": [
            427.15,
            44.801,
          ],
          "v3": [
            349.818,
            44.801,
          ],
        },
        {
          "arcLengthDivisions": 200,
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "LineCurve",
          "v1": [
            349.818,
            44.801,
          ],
          "v2": [
            349.82,
            44.801,
          ],
        },
      ],
      "holes": [
        {
          "arcLengthDivisions": 200,
          "autoClose": false,
          "currentPoint": [
            0,
            0,
          ],
          "curves": [
            {
              "arcLengthDivisions": 200,
              "metadata": {
                "generator": "Curve.toJSON",
                "type": "Curve",
                "version": 4.5,
              },
              "type": "CubicBezierCurve",
              "v0": [
                349.996,
                117.602,
              ],
              "v1": [
                387.113,
                117.602,
              ],
              "v2": [
                417.195,
                147.637,
              ],
              "v3": [
                417.195,
                184.801,
              ],
            },
            {
              "arcLengthDivisions": 200,
              "metadata": {
                "generator": "Curve.toJSON",
                "type": "Curve",
                "version": 4.5,
              },
              "type": "CubicBezierCurve",
              "v0": [
                417.195,
                184.801,
              ],
              "v1": [
                417.195,
                221.965,
              ],
              "v2": [
                387.113,
                252,
              ],
              "v3": [
                349.996,
                252,
              ],
            },
            {
              "arcLengthDivisions": 200,
              "metadata": {
                "generator": "Curve.toJSON",
                "type": "Curve",
                "version": 4.5,
              },
              "type": "CubicBezierCurve",
              "v0": [
                349.996,
                252,
              ],
              "v1": [
                312.879,
                252,
              ],
              "v2": [
                282.797,
                221.965,
              ],
              "v3": [
                282.797,
                184.801,
              ],
            },
            {
              "arcLengthDivisions": 200,
              "metadata": {
                "generator": "Curve.toJSON",
                "type": "Curve",
                "version": 4.5,
              },
              "type": "CubicBezierCurve",
              "v0": [
                282.797,
                184.801,
              ],
              "v1": [
                282.797,
                147.637,
              ],
              "v2": [
                312.879,
                117.602,
              ],
              "v3": [
                349.996,
                117.602,
              ],
            },
            {
              "arcLengthDivisions": 200,
              "metadata": {
                "generator": "Curve.toJSON",
                "type": "Curve",
                "version": 4.5,
              },
              "type": "LineCurve",
              "v1": [
                349.996,
                117.602,
              ],
              "v2": [
                349.996,
                117.602,
              ],
            },
          ],
          "metadata": {
            "generator": "Curve.toJSON",
            "type": "Curve",
            "version": 4.5,
          },
          "type": "Path",
        },
      ],
      "type": "Shape",
      "uuid": undefined,
    },
  ],
}
`;
